### PR TITLE
lsp-rust: support rust-analyzer.showReference lens

### PR DESCRIPTION
**Pull request by nbfalcon**
Add a rust-analyzer specific :action-handler for the showReference lens,
which leverages `lsp-show-xrefs'. Refactor: eliminate
`lsp-execute-command', because it caused the :title to be lost, which is
needed to distinguish between the "references" or "implementations"
variants of that lens.

Defining a new `lsp-interface' to destructure
"rust-analyzer.showReference" was impossible, as the former doesn't
support destructuring lists (the :arguments? paramter is a special,
3-element list.).

### CODE REVIEW

**Comment by yyoncho** [comment]
Please revert the lsp-execute-command elimination because it is used multiple times in different projects (e. g. lsp-java, ccls, etc).

File: clients/lsp-rust.el
Comment (kiennq) [inline clients/lsp-rust.el:591]: This should have different indent or should be on the same line with the function name

File: clients/lsp-rust.el
Comment (yyoncho) [inline clients/lsp-rust.el:589]: lsp-interface is for defining destructuring for objects, not for arrays. dash has destructuring for arrays/lists thus there is no need for lsp-interface to handle that. Thus, this comment can be deleted.

File: clients/lsp-rust.el
Comment (nbfalcon) [reply clients/lsp-rust.el:591]: Then the line would be > 80 columns. Is that preferred?

File: clients/lsp-rust.el
Comment (nbfalcon) [reply clients/lsp-rust.el:589]: Will do that.

**Comment by nbfalcon** [comment]
@yyoncho can I perhaps instead refactor the other projects to not use it? That function needs to be at least be refactored, as it removes the :title field (without it, it is impossible to know whether to use `xref--show-defs' or `xref--show-references'), and it would become a liability in the future, if more fields would be added to the Lens lsp structure, yielding confusing errors because fields become nil out of nowhere.

**Comment by yyoncho** [comment]
> @yyoncho can I perhaps instead refactor the other projects to not use it? That function needs to be at least be refactored, as it removes the :title field (without it, it is impossible to know whether to use `xref--show-defs' or `xref--show-references'), and it would become a liability in the future, if more fields would be added to the Lens lsp structure, yielding confusing errors because fields become nil out of nowhere.

I am fine with this approach.

**Comment by nbfalcon** [comment]
@yyoncho can you give me a list of projects that use it?

**Comment by nbfalcon** [comment]
lsp-java is done.

**Comment by nbfalcon** [comment]
ccls.el is done.

**Comment by yyoncho** [comment]
> @yyoncho can you give me a list of projects that use it?

IDK - you have to check all packages that depend on lsp-mode.

**Comment by kiennq** [comment]
Instead of just deleting code, I think we should probably make that function become obsoleted, and promote `lsp--execute-command` to public function, that ways we will not break any package that depends on `lsp-mode`.
The parameter for the old function `lsp-execute-command` is clear, it will just passing `command` and `arguments`, and not react to anything else, so the user of that API should aware that already.

**Comment by nbfalcon** [comment]
@kiennq actually, `lsp-execute-command` isn't a function that the user calls, it is the other way around: it is a `cl-defgeneric`, which some language server clients provide, instead of using :action-handlers. The problem here is that lens resolution goes trough that function, causing the :title? field to be dropped. This makes it impossible to implement this lens properly, but much worse than that might become a huge blocker in the future. I could theoretically extend `lsp--execute-command` to try to see if a `cl-defmethod` exists, but doing so would involve horrible hacks, and I'd much rather fix all downstream packages instead.

**Comment by nbfalcon** [comment]
Also, from the packages I checked, there are no more uses of it. We'll need to wait for ccls and lsp-java, and then this can be merged.

File: clients/lsp-rust.el
Comment (kiennq) [reply clients/lsp-rust.el:591]: You can put multi-line for the parameters right?
That would make it easier to distinguished between params and body of a function

**Comment by kiennq** [comment]
> actually, `lsp-execute-command` isn't a function that the user calls, it's the other way around: that function is a `cl-defgeneric`, provided by some language server clients. The problem here is that lens resolution goes trough that function, causing the :title? field to be dropped.

Lens resolution is not using `:title` information until now, and for this feature we better switch to another function that also passing the `:title` information, I'm totally agree with that.
But, that's not the reason to delete this function and its functionality.

The `cl-defgeneric` of this function provides an easy way to add new handlers for `:action-handlers` without modifying the `lsp-mode` source code that user can easily extend.

A [quick search](https://github.com/search?q=lsp-execute-command&type=code) of `lsp-execute-command` shows over 1K results, I assume all will be broken (not including private repo) when this change is merged.

**Comment by nbfalcon** [comment]
We cannot use a new function for this. Also, my problem isn't only the :title field, even though it hints the problems that will follow; `lsp-execute-command` was misdesigned: it should pass the entire action trough, not only specific fields of it. Not doing that will become a problem when new fields are added, perhaps even server-specific ones. This API is a liability and needs to be eliminated.

Regarding your github search, most of them are elpa directories containing `lsp-java-boot`. I already made a PR to fix that there. I'll investigate the interesting ones. Your search will be a good starting point, thank you.

**Comment by nbfalcon** [comment]
Nevermind, it's still too many. I propose an alternative: `lsp-execute-command` will still be available as a function, but `cl-defmethod` overrides of it will no longer be possible. This means nothing should be broken.

**Comment by nbfalcon** [comment]
I'll re-add it as a normal defun. It is not overriden using `cl-defmethod` anywhere I didn't provide a PR to fix that though, see [this query]
(https://github.com/search?p=3&q=cl-defmethod+lsp-execute-command+-filename%3Alsp-java-boot.el+-filename%3Accls.el+-filename%3Alsp-java.el+-filename%3Alsp-java.elc+-filename%3Alsp-java-boot.elc+-filename%3Accls.elc+-filename%3A*.elc+-filename%3Alsp-rust.el+-filename%3Alsp-mode.el&type=Code)

Note the exclusions: I have provided PRs for `lsp-java-*` and `ccls.el`. All the results are just copies of `lsp-mode`, ccls.el and lsp-java.el, also in compiled form.

The following PRs need to be merged before this can be, though:
- [lsp-java](https://github.com/emacs-lsp/lsp-java/pull/291)
- [ccls.el](https://github.com/MaskRay/emacs-ccls/pull/104)

File: lsp-mode.el
Comment (yyoncho) [inline lsp-mode.el:5167]: It seems to be that this does not improve backward compatibility

Comment (nbfalcon) [reply lsp-mode.el:5167]: It does. Many places call it as a function. I didn't actually find any other unfixed place that uses `cl-defmethod` on this across the entirety of github, only lots of elpa directories containing outdated ccls.elc? and lsp-java-*.elc? files.

Comment (yyoncho) [reply lsp-mode.el:5167]: ok, this will work, but this is not the usecase covered by this method. It was an extension point, not an entry point supposed to be called.

Comment (nbfalcon) [reply lsp-mode.el:5167]: That is true, however people abused it like that anyway. Do you consider it acceptable to break all that code? (I actually like breaking changes, but we need to be careful for a project this large) There is no way I can go trough 800 projects to make this slightly more elegant.

Comment (yyoncho) [reply lsp-mode.el:5167]: I think we should either clean it up or make sure it is fully backward compatible.

Comment (nbfalcon) [reply lsp-mode.el:5167]: Now I'm unsure: I can't seem to find anyone really using it: [github search](https://github.com/search?q=lsp-execute-command+-filename%3Accls.el+-filename%3Accls.elc+-filename%3Alsp-*&type=Code)

This kak-lsp stuff seems to be unrelated to lsp-mode, and the one use as a keybinding is wrong anyway, since that function wasn't ever interactive.

Comment (nbfalcon) [reply lsp-mode.el:5167]: @yyoncho by clean up, do you mean remove it? From my latest search, it seems that no one actually uses it.

**Comment by nbfalcon** [comment]
We may want to make a new release for this though, as it would break ccls.el and lsp-java-* for anyone who decides to update `lsp-mode` independently of these other packages.

File: lsp-mode.el
Comment (kiennq) [reply lsp-mode.el:5167]: Can we do this in backward compatible way?

Maybe something like
``` elisp
(defvar lsp--from-internal-call nil)
(make-obsolete 'lsp-execute-command 'new-one "7.2")
(cl-defmethod lsp-execute-command (_server command arguments)
  "Dispatch COMMAND execution."
  (if lsp--from-internal-call
      :obsoleted
    (lsp--execute-command (lsp-make-command :command (symbol-name command)
                                            :arguments? arguments)))

(defun lsp-lens--create-interactive-command (command?)
  ;; ...
  (let ((lsp--from-internal-call t))
    (when (eq (lsp-execute-command server-id
                                     (intern (lsp:command-command command?))
                                     (lsp:command-arguments? command?))
                :obsoleted)
      (lsp--execute-command ...))))
```

Comment (yyoncho) [reply lsp-mode.el:5167]: Either this or binding the full command in `lsp-lens--...` and checking if it is bound in lsp--execute-command. Not sure which one is less ugly.

Comment (nbfalcon) [reply lsp-mode.el:5167]: @yyoncho so you mean something like `fboundp` `lsp-execute-command` and leaving the latter undefined in `lsp-mode.el`? This would almost work, since `cl-defmethod` introduces the function if it doesn't yet exist. However, it would break code that calls `lsp-execute-command` as a function.

Comment (yyoncho) [reply lsp-mode.el:5167]: No. The idea of @kiennq  is to make lsp-execute-command work as it is and still pass the title. The @kiennq's proposal can be achieved with some dynamic stope tricks. I am not 100% it worths the effort.

Comment (nbfalcon) [reply lsp-mode.el:5167]: I'll return it with an obsoletion warning, since I just had the following idea: we can leverage `cl-no-applicable-method`, limiting the amount of ugly hacks. It should still be removed at some point, perhaps in the next major release.

**Comment by nbfalcon** [comment]
`lsp-execute-command` now works as it did previously. I tested this with `ccls` before my PR; however, to get it working I had to do two things: use `lsp-show-xrefs` instead of `xref--show-xrefs` (because the latter takes a function, not a list of xrefs) and to handle symbols in `lsp--send-execute-command` because I was getting errors related to `json-serialize` (wrong type argument `json-value-p` for a symbol). These seem unrelated to the changes though (I'll investigate with master).

**Comment by nbfalcon** [comment]
I tested with master, `ccls`' xref lenses are broken independently of this PR (I have made a PR there which fixes these issues, along with not using `lsp-execute-command`), so I believe that this can be merged.

File: lsp-mode.el
Comment (kiennq) [inline lsp-mode.el:5181]: Why don't you just make `lsp-execute-command` obsolete instead? We will have compile warning about that, and that would be standard way to alert package maintainer about obsolete function.

File: clients/lsp-rust.el
Comment (nbfalcon) [reply clients/lsp-rust.el:591]: The options are:
```elisp
(lsp-defun lsp-rust--analyzer-show-references ((&Command
                                                :title :arguments?
                                                [_uri _filepos references]))
```
or
```elisp
(lsp-defun lsp-rust--analyzer-show-references ((&Command :title :arguments?
                                                         [_uri _filepos
                                                               references]))
```
, both of which require two more lines and are IMHO uglier. What do you think?

File: lsp-mode.el
Comment (nbfalcon) [reply lsp-mode.el]: That somehow slipped my mind; concluded from testing, calling that function doesn't cause warnings, but overriding it using `cl-defmethod` does. I couldn't find anything in the Emacs docs about the obsoletion of generics, though.

File: clients/lsp-rust.el
Comment (nbfalcon) [reply clients/lsp-rust.el]: Fixed.

File: clients/lsp-rust.el
Comment (yyoncho) [inline clients/lsp-rust.el:591]: out of curiosity what are the expected `title` values here?

**Review by yyoncho — COMMENTED** [review]
The code looks good to me. I will proceed with the merge after doing manual testing.

File: clients/lsp-rust.el
Comment (nbfalcon) [reply clients/lsp-rust.el:591]: <any number> reference(s)
<any number> definition(s)
E.g. 2 references, 2 definitions. I'm unsure whether reference can actually be encountered in the wild.

**Comment by nbfalcon** [comment]
There is one issue: it throws warnings because `lsp-execute-command` is called even though it is deprecated.

File: clients/lsp-rust.el
Comment (nbfalcon) [reply clients/lsp-rust.el:591]: As a side note, do we prefer scheme-style `?` or common-lisp-style `p` predicate functions? I.e. `s-contains-p` or `s-contains?`.

**Comment by nbfalcon** [comment]
@yyoncho you may want to hold off merging until [my ccls.el PR](https://github.com/MaskRay/emacs-ccls/pull/104) gets merged, as otherwise users will get inevitable byte-compiler warnings.

File: clients/lsp-rust.el
Comment (yyoncho) [reply clients/lsp-rust.el:591]: > As a side note, do we prefer scheme-style `?` or common-lisp-style `p` predicate functions? I.e. `s-contains-p` or `s-contains?`.

Most of the code is written using scheme-style although this is not enforced 100% during reviews.

**Comment by yyoncho** [comment]
> @yyoncho you may want to hold off merging until [my ccls.el PR](https://github.com/MaskRay/emacs-ccls/pull/104) gets merged, as otherwise users will get inevitable byte-compiler warnings.

Is the ccls pr mandatory?

File: clients/lsp-rust.el
Comment (kiennq) [reply clients/lsp-rust.el:589]: Sorry for late reply, both look fine to me.
The current solution has same indent between args and body and that make it quite hard to distinguish them.

File: lsp-lens.el
Comment (kiennq) [inline lsp-lens.el:148]: Can `command?` be `nil` here?
If so, don't you need to check before passing it to `lsp--execute-command`?

**Comment by nbfalcon** [comment]
@yyoncho I'd argue yes, because `ccls` users would get annoying depreciation warnings if they chose to upgrade `lsp-mode`. We could also not deprecate that interface, but it kind-of is a second class citizen here, so it should probably be removed at some point.

File: lsp-lens.el
Comment (nbfalcon) [reply lsp-lens.el:148]: I'd guess no, because it wasn't possible previously (this wouldn't be a regression caused by this PR):
```emacs-lisp
    (if (functionp (lsp:command-command command?))
        (lsp:command-command command?)
      (lambda ()
        (interactive)
        (lsp-execute-command server-id
                             (intern (lsp:command-command command?))
                             (lsp:command-arguments? command?))))
; (if (functionp (lsp:command-command nil)) ...) -> (functionp nil) -> nil
```
This means that the control flow passed to the other branch, creating a lambda that would try to `intern` nil, which errors.

```emacs-lisp
(lsp:command-command nil) ; => nil
(intern nil) ; => error
```

Comment (nbfalcon) [reply lsp-lens.el:148]: This function is called on line 190 in master (without this PR), and only lenses that have a command are passed to it (`-filter #'lsp:command-command`), so I guess command? cannot be nil. I'll note that in the docstring.

Comment (nbfalcon) [reply lsp-lens.el:150]: Fixed.

Comment (kiennq) [reply lsp-lens.el:150]: The reason I'm asking is because of parameter name `command?` which indicates it can be `nil`, we should change the parameter name too.

Comment (nbfalcon) [reply lsp-lens.el:150]: I wouldn't do that, since it is an Elisp-level change, meaning that this entire thing has to be manually tested again, for something this trivial. See [this comment by @yyoncho](https://github.com/emacs-lsp/lsp-mode/pull/2317#discussion_r520169506). Noting that in the docstring is good enough, IMHO.

Also, the name isn't that bad: `command?` also corresponds to the naming of the `CodeLens` and `CodeAction` `:command?` fields, indicating that this function expects them to be passed in.

**Comment by nbfalcon** [comment]
I would like to propose not deprecating `lsp-execute-command`, given that my `ccls` PR apparently won't be merged too soon. @yyoncho what do you think?

File: clients/lsp-rust.el
Comment (nbfalcon) [reply clients/lsp-rust.el:589]: @yyoncho which do you prefer?

Comment (yyoncho) [reply clients/lsp-rust.el:589]: For me all three options are acceptable... Just want to mention that we do not enforce 80 symbols or at least there is no agreement on that ATM.

Comment (nbfalcon) [reply clients/lsp-rust.el:589]: Thank you, I'll keep your latter point in mind. Not changing the call here is the best option then, because it means no new changes have to made and reviewed.

**Comment by yyoncho** [comment]
Can you rebase?

**Review by yyoncho — APPROVED** [review]

**Pull request merged**
